File: sdk/testpilot/src/commonMain/kotlin/co/work/testpilot/Runner.kt
package co.work.testpilot

import co.work.testpilot.utils.EmbeddingElement
import co.work.testpilot.utils.EmbeddingUtils
import com.aallam.openai.api.BetaOpenAI
import com.aallam.openai.api.chat.ChatCompletionRequest
import com.aallam.openai.api.chat.ChatMessage
import com.aallam.openai.api.chat.ChatRole
import com.aallam.openai.api.completion.CompletionRequest
import com.aallam.openai.api.embedding.EmbeddingRequest
import com.aallam.openai.api.model.Model
import com.aallam.openai.api.model.ModelId
import com.aallam.openai.client.Models
import com.aallam.openai.client.OpenAI
import io.ktor.client.*
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json

private enum class OpenAIModel(val idString: String) {
    GPT3_TextEmbeddingAda002("text-embedding-ada-002"),
    GPT3_TextDavinci003("text-davinci-003"),
    GPT4("gpt-4"),
    GPT4_0314("gpt-4-0314"),
}

class Runner(val config: Config) {
    val aiClient = OpenAI(token = config.apiKey)
    val serializer = Json { ignoreUnknownKeys = true }

    // TODO: add @Throws
    @OptIn(BetaOpenAI::class)
    suspend fun getCompletionResponse(ui: String, last: String?, objective: String): String? {
        val response = aiClient.chatCompletion(
            ChatCompletionRequest(
                model = ModelId(OpenAIModel.GPT4_0314.idString),
                messages = listOf(
                    ChatMessage(ChatRole.System, Prompts.system(objective)),
                    ChatMessage(ChatRole.User, Prompts.uiState(last = last, ui = ui)),
                ),
                temperature = config.temperature,
                n = 1,
                maxTokens = config.maxTokens,
            )
        )

        return response.choices.first().message?.content
    }

    // TODO: add @Throws
    @OptIn(BetaOpenAI::class)
    suspend fun splitIntoSteps(objective: String): List<String> {
        val response = aiClient.completion(
            CompletionRequest(
                model = ModelId(OpenAIModel.GPT3_TextDavinci003.idString),
                prompt = Prompts.stepsCompletion(objective),
                maxTokens = config.maxTokens,
                temperature = config.temperature,
                n = 1,
            )
        )

        return serializer.decodeFromString(response.choices.first().text)
    }

    // TODO: add @Throws
    suspend fun searchEmbeddings(input: String, query: String, n: Int = 1): List<String> {
        val texts = input
            .split("\n")
            .filter { it.isNotBlank() }

        val response = aiClient.embeddings(
            EmbeddingRequest(
                model = ModelId(OpenAIModel.GPT3_TextEmbeddingAda002.idString),
                input = texts + listOf(query),
            )
        )
        return EmbeddingUtils.search(
            document = response.embeddings
                .dropLast(1)
                .mapIndexed { index, embedding -> EmbeddingElement(embedding, texts[index]) },
            query = response.embeddings.last(),
            n = n,
        )
    }
}

File: Sources/TestPilotKit/Runner.swift
//
//  Runner.swift
//  TestPilotKit
//
//  Created by Flávio Caetano on 3/13/23.
//

import Foundation
import XCTest
import OpenAIKit
import AsyncHTTPClient

class Runner {
    let httpClient = HTTPClient(eventLoopGroupProvider: .createNew)
    let aiClient: OpenAIKit.Client
    let config: Config
    
    init(config: Config) {
        self.config = config
        self.aiClient = OpenAIKit.Client(
            httpClient: httpClient,
            configuration: config.openAIConfig
        )
    }
}

extension Runner {
    func getCompletionResponse(for ui: String, last: String?, objective: String) async throws -> String? {
        let response = try await aiClient.chats.create(
            model: Model.GPT4.gpt40314,
            messages: [
                .system(content: Prompts.system(objective: objective)),
                .user(content: """
                LAST: \(last ?? "null")
                UI:
                \(ui)
                ---
                YOU:
                """),
            ],
            temperature: config.temperature,
            n: 1,
            maxTokens: config.maxTokens
        )
        print("=================Objective: \(objective)")
        print("=================Completion Response: \(response.choices[0].message.content)")
        return response.choices[0].message.content
    }

    func splitIntoSteps(objective: String) async throws -> [String] {
        let response = try await aiClient.completions.create(
            model: Model.GPT3.textDavinci003,
            prompts: [
                Prompts.stepsCompletion(objective: objective)
            ],
            maxTokens: config.maxTokens,
            temperature: config.temperature,
            n: 1
        )
        
        print("===========Spliting into steps: \(response.choices)")

        return try JSONDecoder().decode([String].self, from: Data(response.choices[0].text.utf8))
    }

    func searchEmbeddings(input: String, query: String, n: Int = 1) async throws -> [String] {
        let texts = input
            .split(separator: "\n")
            .map(String.init)
            .filter { !$0.isEmpty }

        let response = try await aiClient.embeddings.create(input: texts + [query])
            
        return Embedding.search(
            on: response.data
                .dropLast(1)
                .enumerated()
                .map { idx, elem in
                    (elem, texts[idx])
                },
            query: response.data.last!,
            n: n
        )
    }
}

private extension Chat.Message {
    var content: String {
        switch self {
        case let .user(content: result): return result
        case let .system(content: result): return result
        case let .assistant(content: result): return result
        }
    }
}
